services/src/main/java/org/keycloak/authentication/authenticators/x509/AbstractX509ClientCertificateAuthenticator.java

```diff
@@ -145,13 +161,24 @@ static UserIdentityExtractor fromConfig(X509AuthenticatorConfigModel config) {
                     extractor = UserIdentityExtractor.getPatternIdentityExtractor(pattern, func);
                     break;
                 case ISSUERDN:
-                    func = config.isCanonicalDnEnabled() ?
-                        certs -> certs[0].getIssuerX500Principal().getName(X500Principal.CANONICAL) :
-                        certs -> certs[0].getIssuerDN().getName();
-                    extractor = UserIdentityExtractor.getPatternIdentityExtractor(pattern, func);
+                    extractor = UserIdentityExtractor.getPatternIdentityExtractor(pattern, getIssuerDNFunc(config));
                     break;
                 case SERIALNUMBER:
-                    extractor = UserIdentityExtractor.getPatternIdentityExtractor(DEFAULT_MATCH_ALL_EXPRESSION, certs -> certs[0].getSerialNumber().toString());
+                    extractor = UserIdentityExtractor.getPatternIdentityExtractor(DEFAULT_MATCH_ALL_EXPRESSION, getSerialnumberFunc(config));
+                    break;
+                case SHA256_THUMBPRINT:
+                    extractor = UserIdentityExtractor.getPatternIdentityExtractor(DEFAULT_MATCH_ALL_EXPRESSION, certs -> {
+                        try {
+                            return Hashing.sha256().hashBytes(certs[0].getEncoded()).toString().toLowerCase();
```

Comment:
We need sha256 digest represented as String.
HashUtils.hash returns byte[] and additional processing is needed.
Since we're already using bouncycastle, I've tried to combine those two and eliminate google-commons.